en/docs/apis/advanced-topics/connect-to-an-external-identity-provider.md

```diff
@@ -0,0 +1,146 @@
+# Connect to an External Identity Provider
+
+Choreo uses an in-built Identity Provider (IdP) by default to manage OAuth clients and generate tokens required to authenticate Choreo APIs. Choreo also grants organization administrator privileges to users to configure a required authorization server as an identity provider(IdP) via the Choreo Console. This allows you to [add one or more external identity providers](#add-an-external-identity-provider) to your Choreo organization and [obtain access tokens from the identity providers](#obtain-an-access-token-from-an-external-identity-provider) to exchange and get Choreo access tokens to invoke APIs. This capability lets you expose your APIs to users who are external to Choreo.
+
+## Token exchange
+
+Choreo supports the token exchange grant type to exchange a JSON web token (JWT) issued by an external IdP for a Choreo token. The token exchange grant type uses the protocol defined in the [OAuth 2.0 token exchange specification](https://datatracker.ietf.org/doc/html/rfc8693), which describes how to request and obtain security tokens from OAuth 2.0 authorization servers and exchange a security token for another type of token. The following diagram depicts the token exchange flow in Choreo:
+
+![Token exchange flow](../../assets/img/apis/advanced-topics/external-identity-providers/token-exchange-flow.png){.cInlineImage-full}
+
+When you provide a JWT access token (referred to as the `subject_token` in the preceding diagram), a request is sent to the Choreo token endpoint to exchange the JWT with a Choreo JWT. On validation of the request sent, the corresponding IdP configuration is retrieved based on the issuer. Then the `subject_token` validation (signature and expiry time validation) takes place. Successful validation generates a Choreo access token and returns it to you.
```

Comment:
@wasuradananjith, need your feedback on comments 1, 4, 5, 6, and 7 too please.

Reply:
> are we referring to the token generated by Choreo as Choreo JWT? if so, shall we use the term at all places?

+1

Reply:
> Successful validation generates a Choreo access token and returns it to you -> the token generation happens after validation right? to me, it sounds like it's a part of token validation.

No, we are talking about the token exchange here. It means that the user provides a token, and it will be validated inside our system. And upon successful validation, another token will be generated (which is the Choreo JWT) and will be returned to the user. When invoking the APIs in Choreo, the user can only use the Choreo JWT that he got, not the original token provided.

Can you please consider this explanation and change the doc accordingly if needed? This is the simple meaning of token exchange.

Reply:
> When you provide a JWT access token (referred to as the `subject_token` in the preceding diagram), a request is sent to the Choreo token endpoint to exchange the JWT with a Choreo JWT - > should we say it like this? To exchange a JWT to a Choreo JWT, you need to make a request to the Choreo token endpoint including the JWT access token (referred to as the `subject_token` in the preceding diagram) in the request body.

+1

Reply:
> We have missed the authentication part. Is it intentional?

I am not sure whether it is needed to mention. @vithu30 @Nirothipan WDYT?

Reply:
> based on the issuer? -> using the issuer

+1, it makes more sense

Reply:
We have missed the authentication part. Is it intentional >> yes we can leave it as it is of too much details. wdyt @shaniR

Reply:
@Nirothipan, Well we can work around the words and make it a little shorter. I think we are leaving out an important step , while explaining the diagram,

Reply:
As per the discussion had with @shaniR and @Nirothipan , let's mention something like "Upon successful authentication... " and continue our story.

Reply:
@Nashaath  let's just add a few words mentioning something like "upon successful authentication"

Reply:
Applied suggestions based on comments.